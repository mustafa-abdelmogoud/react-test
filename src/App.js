import React, { useState } from "react";
import "./App.css";

function App() {
  const [counter, setCounter] = useState(0);
  return (
    <div className="App">
      <h1 data-testid="counter">{counter}</h1>
      <button data-testid="increment" onClick={() => setCounter(counter + 1)}>
        +
      </button>
      <button data-testid="decrement" onClick={() => setCounter(counter - 1)}>
        -
      </button>
    </div>
  );
}

export default App;
