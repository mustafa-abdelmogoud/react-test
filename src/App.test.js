import React from "react";
import { render, fireEvent, getByTestId } from "@testing-library/react";
import App from "./App";

/**
 * Behaviour
 */
test("initial counter value is 0", () => {
  const { container } = render(<App />);
  /**
   * select hte right query
   * https://testing-library.com/docs/guide-which-query
   */
  const counterValue = getByTestId(container, "counter");
  expect(counterValue.textContent).toBe("0");
});

test("increment by clicking on increment button", () => {
  const { container } = render(<App />);
  const counterValue = getByTestId(container, "counter");
  const incrementButton = getByTestId(container, "increment");
  expect(counterValue.textContent).toBe("0");
  fireEvent.click(incrementButton);
  expect(counterValue.textContent).toBe("1");
});

test("decrement by clicking on increment button", () => {
  const { container } = render(<App />);
  const counterValue = getByTestId(container, "counter");
  const decrementButton = getByTestId(container, "decrement");
  expect(counterValue.textContent).toBe("0");
  fireEvent.click(decrementButton);
  expect(counterValue.textContent).toBe("-1");
});

test("decrement & increment", () => {
  const { container } = render(<App />);
  const counterValue = getByTestId(container, "counter");
  const decrementButton = getByTestId(container, "decrement");
  const incrementButton = getByTestId(container, "increment");
  expect(counterValue.textContent).toBe("0");
  fireEvent.click(incrementButton);
  expect(counterValue.textContent).toBe("1");
  fireEvent.click(decrementButton);
  expect(counterValue.textContent).toBe("0");
});

/**
 * Appearance
 */

// don't test implementation details
test("page shows increment & decrement buttons", () => {
  const { getByText } = render(<App />);
  expect(getByText(/\+/i)).toBeInTheDocument();

  expect(getByText(/-/i)).toBeInTheDocument();
});

// don't run into updating the snapshot
test("match a snapshot", () => {
  const { container } = render(<App />);

  expect(container).toMatchSnapshot();
});
