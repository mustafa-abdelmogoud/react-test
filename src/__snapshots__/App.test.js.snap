// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match a snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <h1
      data-testid="counter"
    >
      0
    </h1>
    <button
      data-testid="increment"
    >
      +
    </button>
    <button
      data-testid="decrement"
    >
      -
    </button>
  </div>
</div>
`;
